File: lib/core/utils/sms_parser/engines/categorization_system.dart
class CategoryMapping {
  final String category;
  final String subcategory;

  const CategoryMapping(this.category, [this.subcategory = 'General']);
}

class CategorizationSystem {
  static final Map<String, CategoryMapping> _merchantToCategory = {
    'zomato': const CategoryMapping('Food'),
    'swiggy': const CategoryMapping('Food'),
    'starbucks': const CategoryMapping('Food'),
    'mcdonalds': const CategoryMapping('Food'),
    'kfc': const CategoryMapping('Food'),
    'blinkit': const CategoryMapping('Food'),
    'zepto': const CategoryMapping('Food'),
    'bigbasket': const CategoryMapping('Food'),
    'uber': const CategoryMapping('Travel'),
    'ola': const CategoryMapping('Travel'),
    'petrol': const CategoryMapping('Travel'),
    'shell': const CategoryMapping('Travel'),
    'hpcl': const CategoryMapping('Travel'),
    'bpcl': const CategoryMapping('Travel'),
    'amazon': const CategoryMapping('Shopping'),
    'flipkart': const CategoryMapping('Shopping'),
    'netflix': const CategoryMapping('Entertainment'),
    'spotify': const CategoryMapping('Entertainment'),
    'hotstar': const CategoryMapping('Entertainment'),
    'airtel': const CategoryMapping('Bills'),
    'jio': const CategoryMapping('Bills'),
    'vi ': const CategoryMapping('Bills'),
    'bescom': const CategoryMapping('Bills'),
    'atm': const CategoryMapping('Other'),
  };

  static CategoryMapping getMapping(String merchant, String normalizedBody, {String type = 'debit'}) {
    if (type == 'credit') {
      if (normalizedBody.contains('salary')) return const CategoryMapping('Salary', 'General');
      return const CategoryMapping('Other', 'General');
    }

    String merchantLower = merchant.toLowerCase();
    
    for (var entry in _merchantToCategory.entries) {
      final key = entry.key.trim();
      final regex = RegExp(r'\b' + key + r'\b');
      if (regex.hasMatch(merchantLower) || regex.hasMatch(normalizedBody)) {
        return entry.value;
      }
    }

    if (RegExp(r'\batm\b').hasMatch(normalizedBody) || RegExp(r'\bcash\b').hasMatch(normalizedBody)) {
      return const CategoryMapping('Other', 'General');
    }
    
    return const CategoryMapping('Other', 'General');
  }

  static String categorize(String merchant, String normalizedBody, {String type = 'debit'}) {
    return getMapping(merchant, normalizedBody, type: type).category;
  }

  static String categorizeSubcategory(String merchant, String normalizedBody, {String type = 'debit'}) {
    return 'General';
  }
}

File: android/app/src/main/kotlin/com/smart_money_tracker/parser/CategorizationSystem.kt
package com.smart_money_tracker.parser
 
data class CategoryMapping(val category: String, val subcategory: String = "General")

object CategorizationSystem {
    private val merchantToCategory = mapOf(
        "zomato" to CategoryMapping("Food"),
        "swiggy" to CategoryMapping("Food"),
        "starbucks" to CategoryMapping("Food"),
        "mcdonalds" to CategoryMapping("Food"),
        "kfc" to CategoryMapping("Food"),
        "blinkit" to CategoryMapping("Food"),
        "zepto" to CategoryMapping("Food"),
        "bigbasket" to CategoryMapping("Food"),
        "uber" to CategoryMapping("Travel"),
        "ola" to CategoryMapping("Travel"),
        "petrol" to CategoryMapping("Travel"),
        "shell" to CategoryMapping("Travel"),
        "hpcl" to CategoryMapping("Travel"),
        "bpcl" to CategoryMapping("Travel"),
        "amazon" to CategoryMapping("Shopping"),
        "flipkart" to CategoryMapping("Shopping"),
        "netflix" to CategoryMapping("Entertainment"),
        "spotify" to CategoryMapping("Entertainment"),
        "hotstar" to CategoryMapping("Entertainment"),
        "airtel" to CategoryMapping("Bills"),
        "jio" to CategoryMapping("Bills"),
        "vi " to CategoryMapping("Bills"),
        "bescom" to CategoryMapping("Bills"),
        "atm" to CategoryMapping("Other")
    )

    fun getMapping(merchant: String, normalizedBody: String, type: String = "debit"): CategoryMapping {
        if (type == "credit") {
            if (normalizedBody.contains("salary")) return CategoryMapping("Salary", "General")
            return CategoryMapping("Other", "General")
        }

        val merchantLower = merchant.lowercase()
        
        for ((keyStr, value) in merchantToCategory) {
            val key = keyStr.trim()
            val regex = Regex("\\b$key\\b")
            if (regex.containsMatchIn(merchantLower) || regex.containsMatchIn(normalizedBody)) {
                return value
            }
        }

        if (Regex("\\batm\\b").containsMatchIn(normalizedBody) || Regex("\\bcash\\b").containsMatchIn(normalizedBody)) {
            return CategoryMapping("Other", "General")
        }
        
        return CategoryMapping("Other", "General")
    }

    fun categorize(merchant: String, normalizedBody: String, type: String = "debit"): String {
        return getMapping(merchant, normalizedBody, type).category
    }

    fun categorizeSubcategory(merchant: String, normalizedBody: String, type: String = "debit"): String {
        return "General"
    }
}
